docs/cip/README.md

```diff
@@ -37,231 +61,765 @@ Leios achieves its scalability through a decoupled block production and aggregat
+
+#### Leios Mini-Protocols
+
+Leios introduces **five new mini-protocols** to handle the additional message types required for EB distribution, voting, and certificate construction.
+
+<div align="center">
+<a name="table-6"></a>
+
+| **Protocol** | **Purpose** | **Timing Constraint** |
+| :----------: | ----------- | --------------------- |
+| **RbHeaderRelay** | Diffuse RB headers for equivocation detection | Must achieve $\Delta_\text{hdr}$ diffusion |
+| **EbRelay** | Diffuse fresh EBs to enable timely validation | Must reach voters within $L_\text{vote}$ |  
+| **VoteRelay** | Diffuse valid votes for certificate aggregation | Must diffuse within $L_\text{diff}$ |
+| **EbFetch** | Retrieve certified EBs for chain reconstruction | On-demand after certificate inclusion |
+| **TxFetch** | Retrieve referenced transactions for EB validation | Before EB validation completes |
+
+_Table 6: Leios Mini-Protocols_
+
+</div>
+
+These protocols implement freshest-first delivery and cryptographic validation to meet their respective timing constraints while preventing spam and DoS attacks.
+
+##### RbHeaderRelay Mini-Protocol
+
+This protocol diffuses RB headers across the network within $\Delta_\text{hdr}$ slots to enable equivocation detection, distributing only headers (not full blocks) to meet this timing bound. By avoiding full block validation, it ensures that all nodes receive RB headers before voting begins, maintaining awareness of all competing chains and equivocations across the network.
+
+**Mini-Protocol Overview**
+
+RbHeaderRelay is a **pull-based relay protocol** that enables nodes to request and receive RB headers from peers. It implements the Relay mini-protocol pattern used in Ouroboros networks for transaction submission, extended with equivocation-aware rules to ensure all nodes can detect competing chains.
+
+**Key Properties**:
+- **Pull-based design**: Nodes request headers from peers, preventing DoS attacks
+- **Freshest-first delivery**: Prioritizes recent headers over historical ones
+- **Equivocation handling**: Serves at most two headers per (slot, issuer) pair
+- **Cryptographic validation**: Signature and VRF verification before serving
+- **Universal availability**: Headers served regardless of fork preference
+
+##### EbRelay Mini-Protocol  
+
+> [!Warning]
+> **TODO**: Protocol specification
+
+##### VoteRelay Mini-Protocol
+
+> [!Warning]
+> **TODO**: Protocol specification
+
+##### EbFetch Mini-Protocol
+
+> [!Warning]
+> **TODO**: Protocol specification
+
+##### TxFetch Mini-Protocol
+
+> [!Warning]
+> **TODO**: Protocol specification
+
+
+### Incentives
+
+> [!Warning]
+>
+> This section is work in progress.
```

Comment:
There should be an incentive for block producers to put transactions into RBs and EBs fairly. There should be a way for users to predict or choose which block their transactions is going in. One of such ways is to define different fee thresholds for RB and EB. This may need adding new protocol parameters.

Just sorting by fee isn't enough, some users may want a garantee their transaction is either put into RB or declined as soon as possible (e.g. time is crucial for DEXes, Djed).

Reply:
@pagio I think this will work, thanks!

> So, I don't see how marking makes tx inclusion faster.

Marking makes receiving response faster and makes the system more responsive, so if the tx is rejected, the user can update the tx according e.g. to the new currency exchange rate, and send an updated one.